util/fpga/cw_spiflash.py

```diff
@@ -0,0 +1,273 @@
+        """Inits SPIProgrammer with the path of a firmware image and board name."""
+        self._firmware_path = firmware_path
+        self._board = board
+
+    def run(self, fpga):
+        """Programs OpenTitan over the SPI interface of SAM3X/U on CW310/305.
+
+        This implementation has two improvements over the `spiflash` tool under
+        sw/host/spiflash:
+        * Optimizes the happy path by checking the acknowledgement of frame N-1
+          after transmitting frame N instead of transmitting an extra frame to check
+          each frame's acknowledgement.
+        * Sends an all 0xFF frame as the last frame to avoid cases where the last
+          frame gets corrupted and halts the bootstrapping process.
+
+        Args:
+          fpga: CW310/305 to be programmed, a ``cw.capture.targets.CW310/305`` instance.
+        """
+        with _Bootstrap(fpga, self._board) as bootstrap:
+            frame_number = 0
+            flash_offset = 0
+            prev_frame = None
+            print(f'Programming OpenTitan with "{self._firmware_path}"...')
+            with open(self._firmware_path, mode='rb') as fw:
+                # Read fixed-size blocks from the firmware image.
+                # Note: The second argument ``b''`` to ``iter`` below is the sentinel value that
+                # ends the loop, i.e. the value returned by ``fw.read`` at EOF.
+                for payload in iter(
+                        partial(fw.read, _SpiFlashFrame.PAYLOAD_SIZE), b''):
+                    frame = _SpiFlashFrame(frame_number, flash_offset, payload)
+                    # We need to make sure that the previous frame is transmitted correctly before
+                    # proceeding with the next frame.
+                    while True:
+                        # Transmit frame N, receive acknowledgement for frame N-1.
+                        ack = bootstrap.transfer(frame)
+                        if prev_frame and prev_frame.expected_ack != ack:
+                            # When OpenTitan encounters a transmission error or an out of
+                            # sequence frame, it ignores the current frame and sends the ack of
+                            # the last successfully received frame. Since transmission errors
+                            # can occur in both directions, i.e. the previous frame sent by
+                            # SAM3U or the ack sent by OpenTitan, and we can't tell them
+                            # apart, we should re-transmit the previous frame and check the
+                            # ack sent by OpenTitan.
+                            print(f'Error transferring {prev_frame}.')
+                            ack = bootstrap.transfer(prev_frame)
+                            if ack == frame.expected_ack:
+                                # Ack sent by OpenTitan was corrupted. Proceed with the next frame.
+                                break
+                            else:
```

Comment:
This else block seems unnecessary?

Reply:
Just a matter of preference. Yes it doesn't do anything because of `pass`, exists just for clarity because there are multiple things nested.